sdk/search/search-documents/src/searchIndexingBufferedSender.ts

```diff
@@ -377,9 +386,35 @@ export class SearchIndexingBufferedSender<T> {
       this.batchObject = new IndexDocumentsBatch<T>();
       while (actions.length > 0) {
         const actionsToSend = actions.splice(0, this.initialBatchActionCount);
-        await this.submitDocuments(actionsToSend, options);
+        await this.submitDocuments(this.pruneActions(actionsToSend, actions), options);
+      }
+    }
+  }
+
+  private pruneActions(
+    batch: IndexDocumentsAction<T>[],
+    actions: IndexDocumentsAction<T>[]
+  ): IndexDocumentsAction<T>[] {
+    let counter: number = 0;
+    const hashSet: Set<string> = new Set<string>();
+    const resultBatch: IndexDocumentsAction<T>[] = [];
+
+    while (counter < batch.length) {
+      const { __actionType, ...document } = batch[counter];
+      // @ts-ignore
+      const key: string = this.documentKeyRetriever(document);
+
+      if (hashSet.has(key)) {
+        actions.unshift(batch[counter]);
+        batch.splice(counter, 1);
+      } else {
+        resultBatch.push(batch[counter]);
+        hashSet.add(key);
+        counter++;
       }
     }
+
+    return resultBatch;
   }
```

Comment:
What if instead of modifying the input arrays, we returned two arrays (one of the actions to submit in the current batch and another that contains the actions to put back into the queue:

```suggestion
  private pruneActions(
    batch: IndexDocumentsAction<T>[],
  ): { batchToSubmit: IndexDocumentsAction<T>[]; submitLater: IndexDocumentsAction<T>[]}  {

    const hashSet: Set<string> = new Set<string>();
    const resultBatch: IndexDocumentsAction<T>[] = [];
    const pruned: IndexDocumentsAction<T>[] = [];
    
    for (const document of batch) {
       const key = this.documentKeyRetriever(document);
       if (hashSet.has(key)) {
          pruned.push(document);
       } else {
          hashSet.add(key);
          resultBatch.push(document);
       }
    }
    return { batchToSubmit: resultBatch, submitLater: pruned };
  }
```

Then the caller can unshift submitLater onto the front of the actions array.